### ISSUE TITLE
2.0 Issue - _destroy support

### ISSUE THREAD

**Issue by davidjsilva** [Knockout-Contrib/KoGrid#134]
In previous versions of koGrid it supported knockout's _destroy and would not display rows that had _destroy=true.  In 2.0, it does not seem to support this.  I have create a fiddle http://jsfiddle.net/55GKa/16/ which has a delete button column but it doesn't hide the row.  Please let me know if this was intentional and if so, how would I delete a row from the grid.  Please keep in mind I don't want to actually remove the array item from the data observableArray.

**Comment by orneryd** [comment]
Hey David. Totalyl seems reasonable and we will get this implemented shortly. Thanks!

**Comment by davidjsilva** [comment]
So just wanted to follow up on this feature.  Wasn't sure if you had a chance to look into this.  If it's going to be a bit before it's supported I will definitely need to take a hack at it.  I was thinking of getting the magic done in renderedChangeNoGroups function but it was still showing a blank row where the _destroy=true rows would have been, so maybe that's not the right spot for this.  Thoughts?

**Comment by davidjsilva** [comment]
OK so I think I found a good spot for this, well it works for me at least, in the kg.SearchProvider.

``` javascript
grid.filteredData(grid.sortedData().filter(function (item) {
  if (item._destroy) {
    return false;
  }
  ...
}));
```

Let me know if that's what you were thinking.

And while I'm at it, I think there is an issue in this function "if (!self.filterText)" always evaluates to false because filterText is observable.  Maybe should it should be - if (!ft)??  And maybe that if condition needs to be moved in front of 

``` javascript
var field = ko.utils.unwrapObservable(item[self.field]);
var fieldMap = ko.utils.unwrapObservable(item[self.fieldMap[self.field]]);
```

so as to avoid unwrapping the observables if not necessary.

**Comment by orneryd** [comment]
I am going to make an update here shortly to koGrid and will incorporate these changes. However, I am going to put the _destory further upstream in the rowFactory line 91.

``` javascript
            if (!item._destroy) {
                var row;
                if (item.isAggRow) {
                    row = self.buildAggregateRow(item, self.renderedRange.topRow + indx);
                } else {
                    row = self.buildEntityRow(item, self.renderedRange.topRow + indx);
                }
                //add the row to our return array
                rowArr.push(row);
            }
```

Sound good?

**Comment by davidjsilva** [comment]
Thanks for taking a look at this.  Unfortunately that fix didn't work for me.  It looks like the part of the code where this "if (!item._destroy)" was added to isn't executed unless the grid is using Grouping.  But even if I add this if condition to the renderedChangeNoGroups function inside the $.each loop we still have the issue where its displaying blank rows.  See attached photo to see what I mean...

![grid-deleted-rows](https://f.cloud.github.com/assets/1336983/8436/14eed8dc-4474-11e2-8aa9-f4a09ce85256.JPG)

Since the array being used in this $.each is from a "slice" of the renderedRange in grid.filteredData I think it's too late at this point.  That's why I added my fix where I did.  Let me know what you think.

**Comment by orneryd** [comment]
I went ahead and moved the location there. let me know how that turns out

**Comment by davidjsilva** [comment]
So this issue isn't completely fixed because it only hides _destroy'ed rows if searchConditions > 0.  Please see the following fiddle for example http://jsfiddle.net/55GKa/16/.  So to test click the Delete button on any row.  You'll notice that the _destroy column says true, but the row is still visible.  Now open up the Column Menu and filter the rows by something.  Then click the Delete button on the filtered rows and they will disappear correctly.

``` javascript
self.evalFilter = function () {
        if (searchConditions.length === 0)
            grid.filteredData(grid.sortedData.peek());
        else {
            grid.filteredData(grid.sortedData.peek().filter(function (item) {
                if (item._destroy) {
                    return false;
                }
```

The item._destroy condition is only executed when searchConditions > 0 which is not the correct behavior.

**Comment by orneryd** [comment]
we'll get this fixed in the next build. However, your fiddle is strange, Why don't you just do it this way? http://jsfiddle.net/timothyswt/55GKa/23/

$userViewModel is accessible in all the templatable areas.

``` javascript
var myCellTemplate = '<button data-bind="click: function(){ $userViewModel.deleteRow($parent); }">delete</button>';
```

instead of 

``` javascript
var getCellTemplate = function ()
{
    return '<button data-bind="click: function () { ko.dataFor($root.$grid.$root[0]).deleteRow($root); }">delete</button>';
};
```

**Comment by davidjsilva** [comment]
Oh yeah thanks for the heads up.  It was an old fiddle that I just touched up to display this delete issue.  The $userViewModel has been incredibly useful and I've updated all my code to use it, I just forgot to update the fiddle.

**Comment by orneryd** [comment]
this should finally be fixed in 2.1.1
